Feature: Input schema

ThunderSon opened the pull request:
- The schema starts with an array of CRE_IDs. We can make it an object that is indexed by the CRE, didn't feel right.
- Links are objects as we need to be able to call each one of them (requirements, tests, etc.) and fill accordingly.
- Every requirement should have an array of IDs which will be referenced.
- These arrays contain objects of every ID pertaining to that standard, and URI for referencing.

Let me know what you think about this structure 😄

This PR partially takes care of #16 and doesn't close it

ThunderSon committed:
feat: input feed file schema

ThunderSon committed:
chore: exclude schemas file from Jekyll build

northdpole commented on schemas/input-feed.json:
```diff
@@ -0,0 +1,46 @@
+[
+  {
+    "CRE_ID": "<ID in here>",
+    "links": {
```
This tries to both define a CRE and link a CRE with multiple other things that are highly subject to change.
Maybe it's better if this is split into two.

A definiton such as:
` {"CRE":{"description":<string>,"id":<fromatted string>,"tags":[<string>], "requirements":[<string>]}}`
and a link such as 
`{"CRE_LINK":{"CRE_ID":<formated_str>,"name":<ASVS,ISO,PCI,etc>,"id":<ASVS_ID,3.3.1a,etc>,"tags":[<a>,<list>,<of>,<free>,<text>,<tags>]}
`
Each link strictly maps one other thing to one CRE. If more mappings are needed a new object can be written.

The link can then be something like:

`{"CRE_LINK":{"CRE_ID":<formated_str>,"name":"ASVS", "id":"ASVS_ID","tags":[]}
`
or

`{"CRE_LINK":{"CRE_ID":<formated_str>,"name":"WSTG", "id":"4.3.2.1","tags":["tests"]}
`

This way, we(or others) can use tag keywords such as "tests", "controls", "documentation", "automation" to tag almost arbitrary pieces of documentation or functionality.
Then just filtering all "CRE_LINKS" to generate a massive diagram (or mindmap/other graph) of how everything maps together could be automated

ThunderSon replied:
This document pertains to people that are pushing updates to the CREs, and not creating CREs. CRE metadata is not required in our case. This is more focused to let a project to integrate their IDs with the CREs. The links in the object are links to ASVS-ID, WSTG-ID, etc., and not the CRE.

northdpole replied:
i see, so what do you think about the second part of the suggestion(one "CRE_LINK" per json object with tags, instead of one object with an array of CRE links each under some category?

northdpole reviewed (commented):
let me know what you think

ThunderSon committed:
chore: change from JSON to YAML

ThunderSon committed:
chore: name change to data representation

Signed-off-by: Elie Saad <eliesaad7@gmail.com>

ThunderSon commented:
- Changed from JSON to YAML for better readability and to look easier for people to use.
- Made a bigger object `CREs` that contains an array of `CRE` objects.
- Added the metadata object inside it.

Feedback questions:
- What are the thoughts on this whole structure?
- Do we require the `links` object?
- Do you prefer reading an `ID` or a `Name` in the `CRE ID`? _e.g._ `cre_relevant_details` vs `1004`

kingthorin commented on schemas/data.yaml:
```diff
@@ -0,0 +1,28 @@
+CREs:
+  - CRE: "<ID in here>" # ID could be a number, or a keyword. If it's a keyword, an ID will be set in the metadata.
```
Based on the comment this is imprecise perhaps that's intentional and I'm missing the vision?

Also this is going to sound silly, but if I was a member or dev that had never heard of this standard/effort what does the acronym CRE actually mean? (Common Requirement Enumeration? 🤷‍♂️ )

ThunderSon replied:
Can you discuss actions for both points? This is still being built, so I need to see what you'd be aiming at precisely.
1. We still didn't pick what should be the identifier for the CRE -- so this is still not set in stone, trying to take feedback on it and we're delving deeper into which to use as the main identifier.
2. Should we set it as Common_Requirement_Enumeration? CWE is used as CWE and everyone uses it as such. ISO, PCIDSS, ASVS; how is this going to be different?

kingthorin replied:
We should absolutely discuss that's the whole point of this process right?

1. I'm good with either ID or keyword, however keyword may become difficult as it's more qualitative than quantitative (people will have different definitions/interpretations of keywords). At the same time ID(s) will require more maintenance and perhaps a boolean `isDeprecated` (or whatever) field.
2. I'm good with whatever, but it should be included as a comment somewhere for clarity. Honestly I picked "common requirement enumeration' out of thin air. I dunno what you or others were using terminology wise.

ThunderSon replied:
Oh, I am saying that CRE is that, and just replacing it with that. I feel like a comment would do to make it more understandable.
I'll have a deeper look into the `ID` on Monday.
Any points on the structure or does it look good for you?

northdpole replied:
@kingthorin
maybe an id could be structured as:
CATEGORY_ID-Num_in_category
so CRE "always encrypt a database"
could be CRYPTO-2
alternatively, if grouping CREs is confusing, it could be just an id
1234-2
1234 being an arbitrary id.

ThunderSon replied:
The ID naming is something a little bit bigger than this. @northdpole you can see that in the sheet we have. Let's not focus this discussion in here. This PR is not precisely for the ID structure, if you want we can create an issue and get in bigger input.

robvanderveer replied:
There is no need to distinguish between types of sources (requirements vs tests) here, for two reasons: 1) the type can be ambiguous: even though the testing guide is written as a test guide, it also serves well as a requirement document for developers. Look at the ASVS: it is written as a set of checks, so you could also call that a 'test' source. Better not. And 2) IF we want to distinguish between types of sources, this is better stored once at the metadata itself. That makes this type easier to maintain and the CRE content more elegant.

robvanderveer replied:
The ID's of the sources (e.g. ASVS-ID) should be optional and not the URI. The Purpose of the CRE is to refer to the right section. Not to keep a list for reference. Having to keep an ID increases maintenance.

bkimminich replied:
I would assume the IDs are more stable than URIs. The OWASP Top 10 entry ID `A1:2017` stayed the same when its location moved from <https://owasp.org/index.php/Top_10-2017_A1-Injection> to legacy <https://wiki.owasp.org/index.php/Top_10-2017_A1-Injection> and now redirects to <https://owasp.org/www-project-top-ten/OWASP_Top_Ten_2017/Top_10-2017_A1-Injection>. Shouldn't the CREs be more of a "meta list" of well-known requirements than a "link collection"? I'd imagined it like something similar to page 2 of https://www.cisecurity.org/wp-content/uploads/2017/03/Poster_Winter2016_CSCs.pdf just for AppSec. It surely should have the links, just not as the primary identifier for a requirement in my opinion.

robvanderveer replied:
On ID's versus links. There are two reasons for hyperlinks to be the primary pointer to specific content in a standard. 
1) Link-granularity: often you want to link to a specific section in a standard text that does not have its own ID in the standard. For example: in WSTG the part on checking random session ID's is just one of the many paragraphs in the session testing section. You want to point to the right one. In other words: often a CRE topic has no corresponding ID in a source.
2) Self-maintenance: Yes the IDs are more stable and they are very helpful to have in case the hyperlink is outdated - which is why the IDs should be optional at least. At the same time: one of the key design principles of CRE is that it is self-maintaining and therefore that everything has been to make the links work. In the past, every mapping initiative that needed to be maintained by hand  failed. It's a dead end.
In the case of ASVS this should be automatic as the idea is either 1)the CRE-IDs are embedded IN the content of the ASVS or 2)the maintenance process ensures that changes of the CRE-ID mapping are always done when changes in the ASVS take place.
There are other cases where maintenance of CRE in a standard is a separate process and can lack behind. Typically these are standard that are not yet committed to maintaining their CRE links. In that case an ID can be very handy.

robvanderveer replied:
Regarding CIS controls: yes, that's a good comparison. And there are many more comparisons to make: requirements are controls and guidelines, and rules, and criteria, and checks, and verification items. Basically all standards aim for consensus on these 'things'. The CIS controls are quite high level. CRE aims to be much more towards the level of ASVS. Makes sense?

ThunderSon replied:
Following some discussions, it was agreed that the CRE ID will follow the digits convention (_e.g._ CRE XXXX). The number of digits and starting point will be discussed in an upcoming call. If you have any insights on this, please let us know as well :smile:

northdpole committed:
Merge branch 'master' into feat/input-schema
